src/TurbulenceClosures/turbulence_closure_implementations/TKEBasedVerticalDiffusivities/catke_vertical_diffusivity.jl

```diff
@@ -144,6 +144,36 @@ catke_first(catke1::FlavorOfCATKE, catke2::FlavorOfCATKE) = error("Can't have tw
 ##### Diffusivities and diffusivity fields utilities
 #####
 
+struct CATKEDiffusivityFields{K, L, J, T, U, KC, LC}
+    κu :: K
+    κc :: K
+    κe :: K
+    Le :: L
+    Jᵇ :: J
+    previous_compute_time :: T
+    previous_velocities :: U
+    _tupled_tracer_diffusivities :: KC
+    _tupled_implicit_linear_coefficients :: LC
+end
+
+Adapt.adapt_structure(to, catke_diffusivity_fields::CATKEDiffusivityFields) =
+    CATKEDiffusivityFields(adapt(to, catke_diffusivity_fields.κu),
+                           adapt(to, catke_diffusivity_fields.κc),
+                           adapt(to, catke_diffusivity_fields.κe),
+                           adapt(to, catke_diffusivity_fields.Le),
+                           adapt(to, catke_diffusivity_fields.Jᵇ),
+                           previous_compute_time[],
+                           adapt(to, previous_velocities),
+                           adapt(to, _tupled_tracer_diffusivities),
+                           adapt(to, _tupled_implicit_linear_coefficients))
+
+function fill_halo_regions!(catke_diffusivity_fields::CATKEDiffusivityFields, args...; kw...)
+    fill_halo_regions!(catke_diffusivity_fields.κu, args...; kw...)
+    fill_halo_regions!(catke_diffusivity_fields.κc, args...; kw...)
+    fill_halo_regions!(catke_diffusivity_fields.κe, args...; kw...)
+    return nothing
+end
```

Comment:
These fill halos can be tupled up.

Reply:
touche

Reply:
Just to say, the reason not to do this is that we've had some issues differentiating through the tupling code which previously was type unstable. So the whole reason to put this here was to avoid that. Neverthless we should be able to make it work, I'l try